File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/CalendarInterval.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.types

import java.io.Serializable
import java.util.Locale
import java.util.regex.Pattern


/**
 * The internal representation of interval type.
 */

object CalendarInterval {
  val MICROS_PER_MILLI = 1000L
  val MICROS_PER_SECOND: Long = MICROS_PER_MILLI * 1000
  val MICROS_PER_MINUTE: Long = MICROS_PER_SECOND * 60
  val MICROS_PER_HOUR: Long = MICROS_PER_MINUTE * 60
  val MICROS_PER_DAY: Long = MICROS_PER_HOUR * 24
  val MICROS_PER_WEEK: Long = MICROS_PER_DAY * 7

  /**
    * A function to generate regex which matches interval string's unit part like "3 years".
    *
    * First, we can leave out some units in interval string, and we only care about the value of
    * unit, so here we use non-capturing group to wrap the actual regex.
    * At the beginning of the actual regex, we should match spaces before the unit part.
    * Next is the number part, starts with an optional "-" to represent negative value. We use
    * capturing group to wrap this part as we need the value later.
    * Finally is the unit name, ends with an optional "s".
    */
  private def unitRegex(unit: String) = "(?:\\s+(-?\\d+)\\s+" + unit + "s?)?"

  private val p = Pattern.compile("interval" + unitRegex("year") + unitRegex("month") +
    unitRegex("week") + unitRegex("day") + unitRegex("hour") + unitRegex("minute") +
    unitRegex("second") + unitRegex("millisecond") + unitRegex("microsecond"))
  private val yearMonthPattern = Pattern.compile("^(?:['|\"])?([+|-])?(\\d+)-(\\d+)(?:['|\"])?$")
  private val dayTimePattern = Pattern.compile("^(?:['|\"])?([+|-])?((\\d+) )?" +
    "(\\d+):(\\d+):(\\d+)(\\.(\\d+))?(?:['|\"])?$")
  private val quoteTrimPattern = Pattern.compile("^(?:['|\"])?(.*?)(?:['|\"])?$")

  private def toLong(s: String) = if (s == null) 0L
  else s.toLong

  /**
    * Convert a string to CalendarInterval. Return null if the input string is not a valid interval.
    * This method is case-sensitive and all characters in the input string should be in lower case.
    */
  def fromString(s: String): CalendarInterval = {
    if (s == null) return null
    val m = p.matcher(s.trim)
    if (!m.matches || s == "interval") null
    else {
      val months = toLong(m.group(1)) * 12 + toLong(m.group(2))
      var microseconds = toLong(m.group(3)) * MICROS_PER_WEEK
      microseconds += toLong(m.group(4)) * MICROS_PER_DAY
      microseconds += toLong(m.group(5)) * MICROS_PER_HOUR
      microseconds += toLong(m.group(6)) * MICROS_PER_MINUTE
      microseconds += toLong(m.group(7)) * MICROS_PER_SECOND
      microseconds += toLong(m.group(8)) * MICROS_PER_MILLI
      microseconds += toLong(m.group(9))
      new CalendarInterval(months.toInt, microseconds)
    }
  }

  /**
    * Convert a string to CalendarInterval. Unlike fromString, this method is case-insensitive and
    * will throw IllegalArgumentException when the input string is not a valid interval.
    *
    * @throws IllegalArgumentException if the string is not a valid internal.
    */
  def fromCaseInsensitiveString(s: String): CalendarInterval = {
    if (s == null || s.trim.isEmpty) {
      throw new IllegalArgumentException("Interval cannot be null or blank.")
    }
    val sInLowerCase = s.trim.toLowerCase(Locale.ROOT)
    val interval = if (sInLowerCase.startsWith("interval ")) sInLowerCase
    else "interval " + sInLowerCase
    val cal = fromString(interval)
    if (cal == null) throw new IllegalArgumentException("Invalid interval: " + s)
    cal
  }

  @throws[IllegalArgumentException]
  def toLongWithRange(fieldName: String, s: String, minValue: Long, maxValue: Long): Long = {
    var result = 0L
    if (s != null) {
      result = s.toLong
      if (result < minValue || result > maxValue) {
        throw new IllegalArgumentException(
          s"${fieldName} ${result} outside range [${minValue}, ${maxValue}]")
      }
    }
    result
  }

  /**
    * Parse YearMonth string in form: [-]YYYY-MM
    *
    * adapted from HiveIntervalYearMonth.valueOf
    */
  @throws[IllegalArgumentException]
  def fromYearMonthString(s: String): CalendarInterval = {
    var result : CalendarInterval = null
    if (s == null) {
      throw new IllegalArgumentException("Interval year-month string was null")
    }

    val m = yearMonthPattern.matcher(s.trim)
    if (!m.matches) {
      throw new IllegalArgumentException(
        "Interval string does not match year-month format of 'y-m': " + s)
    }
    else try {
      val sign = if (m.group(1) != null && m.group(1) == "-") -1
      else 1
      val years = toLongWithRange("year", m.group(2), 0, Integer.MAX_VALUE).toInt
      val months = toLongWithRange("month", m.group(3), 0, 11).toInt
      result = new CalendarInterval(sign * (years * 12 + months), 0)
    } catch {
      case e: Exception =>
        throw new IllegalArgumentException(
          "Error parsing interval year-month string: " + e.getMessage, e)
    }
    result
  }

  /**
    * Parse dayTime string in form: [-]d HH:mm:ss.nnnnnnnnn and [-]HH:mm:ss.nnnnnnnnn
    *
    * adapted from HiveIntervalDayTime.valueOf
    */
  @throws[IllegalArgumentException]
  def fromDayTimeString(s: String): CalendarInterval = {
    var result : CalendarInterval = null
    if (s == null) throw new IllegalArgumentException("Interval day-time string was null")
    val m = dayTimePattern.matcher(s.trim)
    if (!m.matches) {
      throw new IllegalArgumentException(
        "Interval string does not match day-time format of 'd h:m:s.n': " + s)
    }
    else try {
      val sign = if (m.group(1) != null && m.group(1) == "-") -1
      else 1
      val days = if (m.group(2) == null) 0
      else toLongWithRange("day", m.group(3), 0, Integer.MAX_VALUE)
      val hours = toLongWithRange("hour", m.group(4), 0, 23)
      val minutes = toLongWithRange("minute", m.group(5), 0, 59)
      val seconds = toLongWithRange("second", m.group(6), 0, 59)
      // Hive allow nanosecond precision interval
      val nanos = toLongWithRange("nanosecond", m.group(8), 0L, 999999999L)
      result = new CalendarInterval(0, sign * (days * MICROS_PER_DAY +
        hours * MICROS_PER_HOUR + minutes * MICROS_PER_MINUTE +
        seconds * MICROS_PER_SECOND + nanos / 1000L))
    } catch {
      case e: Exception =>
        throw new IllegalArgumentException(
          "Error parsing interval day-time string: " + e.getMessage, e)
    }
    result
  }

  @throws[IllegalArgumentException]
  def fromSingleUnitString(unit: String, s: String): CalendarInterval = {
    var result : CalendarInterval = null
    if (s == null) {
      throw new IllegalArgumentException(String.format("Interval %s string was null", unit))
    }

    val m = quoteTrimPattern.matcher(s.trim)
    if (!m.matches) {
      throw new IllegalArgumentException(
        "Interval string does not match day-time format of 'd h:m:s.n': " + s)
    }
    else try
      unit match {
        case "year" =>
          val year = toLongWithRange("year", m.group(1),
            Integer.MIN_VALUE / 12, Integer.MAX_VALUE / 12).toInt
          result = new CalendarInterval(year * 12, 0L)
        case "month" =>
          val month = toLongWithRange("month", m.group(1),
            Integer.MIN_VALUE, Integer.MAX_VALUE).toInt
          result = new CalendarInterval(month, 0L)
        case "week" =>
          val week = toLongWithRange("week", m.group(1),
            Long.MinValue / MICROS_PER_WEEK, Long.MaxValue / MICROS_PER_WEEK)
          result = new CalendarInterval(0, week * MICROS_PER_WEEK)
        case "day" =>
          val day = toLongWithRange("day", m.group(1),
            Long.MinValue / MICROS_PER_DAY, Long.MaxValue / MICROS_PER_DAY)
          result = new CalendarInterval(0, day * MICROS_PER_DAY)
        case "hour" =>
          val hour = toLongWithRange("hour", m.group(1),
            Long.MinValue / MICROS_PER_HOUR, Long.MaxValue / MICROS_PER_HOUR)
          result = new CalendarInterval(0, hour * MICROS_PER_HOUR)
        case "minute" =>
          val minute = toLongWithRange("minute", m.group(1),
            Long.MinValue / MICROS_PER_MINUTE, Long.MaxValue / MICROS_PER_MINUTE)
          result = new CalendarInterval(0, minute * MICROS_PER_MINUTE)
        case "second" =>
          val micros = parseSecondNano(m.group(1))
          result = new CalendarInterval(0, micros)
        case "millisecond" =>
          val millisecond = toLongWithRange("millisecond", m.group(1),
            Long.MinValue / MICROS_PER_MILLI, Long.MaxValue / MICROS_PER_MILLI)
          result = new CalendarInterval(0, millisecond * MICROS_PER_MILLI)
        case "microsecond" =>
          val micros = m.group(1).toLong
          result = new CalendarInterval(0, micros)
      }
    catch {
      case e: Exception =>
        throw new IllegalArgumentException("Error parsing interval string: " + e.getMessage, e)
    }
    result
  }

  /**
    * Parse second_nano string in ss.nnnnnnnnn format to microseconds
    */
  @throws[IllegalArgumentException]
  def parseSecondNano(secondNano: String): Long = {
    val parts = secondNano.split("\\.")
    if (parts.length == 1) {
      toLongWithRange("second", parts(0), Long.MinValue/ MICROS_PER_SECOND,
        Long.MaxValue / MICROS_PER_SECOND) * MICROS_PER_SECOND
    }
    else if (parts.length == 2) {
      val seconds = if (parts(0) == "") 0L
      else {
        toLongWithRange("second", parts(0), Long.MinValue / MICROS_PER_SECOND,
          Long.MaxValue / MICROS_PER_SECOND)
      }
      val nanos = toLongWithRange("nanosecond", parts(1), 0L, 999999999L)
      seconds * MICROS_PER_SECOND + nanos / 1000L
    }
    else {
      throw new IllegalArgumentException(
        "Interval string does not match second-nano format of ss.nnnnnnnnn")
    }
  }
}

final class CalendarInterval(val months: Int, val microseconds: Long) extends Serializable {
  def milliseconds: Long = this.microseconds / CalendarInterval.MICROS_PER_MILLI

  def add(that: CalendarInterval): CalendarInterval = {
    val months = this.months + that.months
    val microseconds = this.microseconds + that.microseconds
    new CalendarInterval(months, microseconds)
  }

  def subtract(that: CalendarInterval): CalendarInterval = {
    val months = this.months - that.months
    val microseconds = this.microseconds - that.microseconds
    new CalendarInterval(months, microseconds)
  }

  def negate : CalendarInterval = new CalendarInterval(-this.months, -this.microseconds)

  override def equals(other: Any): Boolean = {
    if (this equals other) return true
    if (other == null || !other.isInstanceOf[CalendarInterval]) return false
    val o = other.asInstanceOf[CalendarInterval]
    this.months == o.months && this.microseconds == o.microseconds
  }

  override def hashCode: Int = 31 * months + microseconds.toInt

  override def toString: String = {
    val sb = new StringBuilder("interval")
    if (months != 0) {
      appendUnit(sb, months / 12, "year")
      appendUnit(sb, months % 12, "month")
    }
    if (microseconds != 0) {
      var rest = microseconds
      appendUnit(sb, rest / CalendarInterval.MICROS_PER_WEEK, "week")
      rest %= CalendarInterval.MICROS_PER_WEEK
      appendUnit(sb, rest / CalendarInterval.MICROS_PER_DAY, "day")
      rest %= CalendarInterval.MICROS_PER_DAY
      appendUnit(sb, rest / CalendarInterval.MICROS_PER_HOUR, "hour")
      rest %= CalendarInterval.MICROS_PER_HOUR
      appendUnit(sb, rest / CalendarInterval.MICROS_PER_MINUTE, "minute")
      rest %= CalendarInterval.MICROS_PER_MINUTE
      appendUnit(sb, rest / CalendarInterval.MICROS_PER_SECOND, "second")
      rest %= CalendarInterval.MICROS_PER_SECOND
      appendUnit(sb, rest / CalendarInterval.MICROS_PER_MILLI, "millisecond")
      rest %= CalendarInterval.MICROS_PER_MILLI
      appendUnit(sb, rest, "microsecond")
    }
    else if (months == 0) sb.append(" 0 microseconds")
    sb.toString
  }

  private def appendUnit(sb: StringBuilder, value: Long, unit: String): Unit = {
    if (value != 0) sb.append(' ').append(value).append(' ').append(unit).append('s')
  }
}

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.unsafe.types;

import org.apache.spark.annotation.Unstable;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.Duration;
import java.time.Period;
import java.time.temporal.ChronoUnit;
import java.util.Objects;

import static org.apache.spark.sql.catalyst.util.DateTimeConstants.*;

/**
 * The class representing calendar intervals. The calendar interval is stored internally in
 * three components:
 * <ul>
 *   <li>an integer value representing the number of `months` in this interval,</li>
 *   <li>an integer value representing the number of `days` in this interval,</li>
 *   <li>a long value representing the number of `microseconds` in this interval.</li>
 * </ul>
 *
 * The `months` and `days` are not units of time with a constant length (unlike hours, seconds), so
 * they are two separated fields from microseconds. One month may be equal to 28, 29, 30 or 31 days
 * and one day may be equal to 23, 24 or 25 hours (daylight saving).
 *
 * @since 3.0.0
 */
@Unstable
public final class CalendarInterval implements Serializable, Comparable<CalendarInterval> {
  // NOTE: If you're moving or renaming this file, you should also update Unidoc configuration
  // specified in 'SparkBuild.scala'.
  public final int months;
  public final int days;
  public final long microseconds;

  // CalendarInterval is represented by months, days and microseconds. Months and days are not
  // units of time with a constant length (unlike hours, seconds), so they are two separated fields
  // from microseconds. One month may be equal to 29, 30 or 31 days and one day may be equal to
  // 23, 24 or 25 hours (daylight saving)
  public CalendarInterval(int months, int days, long microseconds) {
    this.months = months;
    this.days = days;
    this.microseconds = microseconds;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    CalendarInterval that = (CalendarInterval) o;
    return months == that.months &&
      days == that.days &&
      microseconds == that.microseconds;
  }

  @Override
  public int hashCode() {
    return Objects.hash(months, days, microseconds);
  }

  @Override
  public String toString() {
    if (months == 0 && days == 0 && microseconds == 0) {
      return "0 seconds";
    }

    StringBuilder sb = new StringBuilder();

    if (months != 0) {
      appendUnit(sb, months / 12, "years");
      appendUnit(sb, months % 12, "months");
    }

    appendUnit(sb, days, "days");

    if (microseconds != 0) {
      long rest = microseconds;
      appendUnit(sb, rest / MICROS_PER_HOUR, "hours");
      rest %= MICROS_PER_HOUR;
      appendUnit(sb, rest / MICROS_PER_MINUTE, "minutes");
      rest %= MICROS_PER_MINUTE;
      if (rest != 0) {
        String s = BigDecimal.valueOf(rest, 6).stripTrailingZeros().toPlainString();
        sb.append(s).append(" seconds ");
      }
    }

    sb.setLength(sb.length() - 1);
    return sb.toString();
  }

  private void appendUnit(StringBuilder sb, long value, String unit) {
    if (value != 0) {
      sb.append(value).append(' ').append(unit).append(' ');
    }
  }

  /**
   * Extracts the date part of the interval.
   * @return an instance of {@code java.time.Period} based on the months and days fields
   *         of the given interval, not null.
   */
  public Period extractAsPeriod() { return Period.of(0, months, days); }

  /**
   * Extracts the time part of the interval.
   * @return an instance of {@code java.time.Duration} based on the microseconds field
   *         of the given interval, not null.
   * @throws ArithmeticException if a numeric overflow occurs
   */
  public Duration extractAsDuration() { return Duration.of(microseconds, ChronoUnit.MICROS); }

  /**
   * This method is not used to order CalendarInterval instances, as they are not orderable and
   * cannot be used in a ORDER BY statement.
   * Instead, it is used to find identical interval instances for aggregation purposes.
   * It compares the 'months', 'days', and 'microseconds' fields of this CalendarInterval
   * with another instance. The comparison is done first on the 'months', then on the 'days',
   * and finally on the 'microseconds'.
   *
   * @param o The CalendarInterval instance to compare with.
   * @return Zero if this object is equal to the specified object, and non-zero otherwise
   */
  @Override
  public int compareTo(CalendarInterval o) {
    if (this.months != o.months) {
      return Integer.compare(this.months, o.months);
    } else if (this.days != o.days) {
      return Integer.compare(this.days, o.days);
    } else {
      return Long.compare(this.microseconds, o.microseconds);
    }
  }
}
